sql/wsrep_trans_observer.h

```diff
@@ -144,6 +145,49 @@ static inline int wsrep_start_trx_if_not_started(THD* thd)
   return ret;
 }
 
+/*
+  Helper method to determine if a transaction must terminate
+  a prepared XA by XID (possibly on behalf of another node).
+
+  Return true if the caller must call wsrep_commit_or_rollback_by_xid,
+  false otherwise.
+ */
+static inline bool wsrep_must_commit_or_rollback_by_xid(THD* thd)
+{
+  bool ret= false;
+  if (!thd->wsrep_applier)
+  {
+    char xid_buf[SQL_XIDSIZE];
+    const uint xid_len(get_sql_xid(thd->lex->xid, xid_buf));
+    const std::string xid_string(xid_buf, xid_len);
+    Wsrep_server_state& server_state(Wsrep_server_state::instance());
+    wsrep::high_priority_service* sa(
+      server_state.find_streaming_applier(xid_string));
+    if (sa)
+    {
+      ret= (sa->transaction().state() == wsrep::transaction::s_prepared);
+    }
+  }
+  return ret;
+}
+
+/*
+  Commits or rolls back transaction with the given xid.
+  This is used in the case where prepared XA transaction
+  is terminated on behalf of another node which failed or
+  went non-Primary.
```

Comment:
There is also the case of the client/session (TM) crashing after prepare. Mention it here?